### PR TITLE
containerd: Image delete fixes and cleanup

### PR DESCRIPTION
**Pull request by dmcgowan**
- addresses various tests in https://github.com/moby/moby/issues/46760

Ensure that when removing an image, an image is checked consistently against the images with the same target digest.
Update logic of determining which references are the same to fix failing tests.
Add unit testing around delete.

### CODE REVIEW

File: daemon/containerd/image_delete.go [hunk 1]
```diff
@@ -205,19 +239,61 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
 	})
 }
 
-// isSingleReference returns true if there are no other images in the
+// getSameReferences returns true if there are no other images in the
```
Comment (rumpl) [inline daemon/containerd/image_delete.go:242]: This method doesn't return a boolean any more, could you update the docs?

File: daemon/containerd/image_delete.go [hunk 2]
```diff
@@ -205,19 +239,61 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
...
+				} else {
+					if digestRefs != nil {
+						digestRefs = append(digestRefs, ref)
+					}
+					// Add digest refs at end if no other tags in the same name
+					continue
+				}
+			} else {
+				// Ignore names which do not parse
+				log.G(ctx).WithError(err).WithField("image", ref.Name).Info("failed to parse image name, ignoring")
```
Comment (vvoland) [inline daemon/containerd/image_delete.go:288]: Use `WithFields`?

File: daemon/containerd/image_delete_test.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+				containers:    emptyTestContainerStore(),
+				eventsService: daemonevents.New(),
+			}
+			for _, img := range tc.starting {
+				if _, err := service.images.Create(ctx, img); err != nil {
+					t.Fatalf("failed to create image %q: %v", img.Name, err)
+				}
+			}
+
+			//func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]imagetypes.DeleteResponse, error) {
```
Comment (vvoland) [inline daemon/containerd/image_delete_test.go:168]: Do we need this comment?

File: daemon/containerd/soft_delete.go [hunk 4]
```diff
@@ -17,23 +17,18 @@ const imageNameDanglingPrefix = "moby-dangling@"
...
-
+func (i *ImageService) softImageDelete(ctx context.Context, img *containerdimages.Image, imgs []containerdimages.Image) error {
 	// From this point explicitly ignore the passed context
 	// and don't allow to interrupt operation in the middle.
 
+	// If no explicit image to delete, just return and keep remaining refereneces
+	if img == nil {
+		return nil
+	}
+
```
Comment (vvoland) [inline daemon/containerd/soft_delete.go:28]: If nil img doesn't do anything, maybe we should just accept non nilable `containerdimages.Image`?

**Review by vvoland — COMMENTED** [review]
Just posting a couple of comments/nits.

File: daemon/containerd/image_delete.go [hunk 5]
```diff
@@ -53,123 +53,157 @@ import (
...
 		}
-		return nil
-	})
-	if err != nil {
-		return nil, err
 	}
 	defer func() {
-		if err := i.unleaseSnapshotsFromDeletedConfigs(compatcontext.WithoutCancel(ctx), possiblyDeletedConfigs); err != nil {
-			log.G(ctx).WithError(err).Warn("failed to unlease snapshots")
+		// TODO(dmcgowan): Consider leasing configs directly rather than snapshots to make this unnecessary...
```
[the remark this answers is not in the document]
Comment (dmcgowan) [reply daemon/containerd/image_delete.go:176]: Should I update the comment here for that?

File: daemon/containerd/image_delete.go (hunk 2 above)
Comment (dmcgowan) [reply daemon/containerd/image_delete.go:295]: I don't usually use `WithFields` for errors and the error key is not aliased in the `log` package. There might have been a place I used `WithField` twice and should update to `WithFields` though.

File: daemon/containerd/image_delete.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vvoland) [reply daemon/containerd/image_delete.go:176]: Opened a ticket: https://github.com/moby/moby/issues/46858

I think you can remove that TODO completely now.

File: daemon/containerd/image_test.go [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+		lookup string
+		img    *images.Image
+		all    []images.Image
+		err    error
+	}{
+		{
+			// Get ubuntu images with default "latest" tag
+			lookup: "ubuntu",
+			img:    &testImages[0],
+			all:    []images.Image{testImages[0], testImages[1]},
```
Comment (vvoland) [inline daemon/containerd/image_test.go:83]: Instead of referencing the test image by their index, could we name them?

```go
ubuntuLatest :=  images.Image{
	Name:   "docker.io/library/ubuntu:latest",
	Target: desc(10),
}
ubuntuLatestDigested :=  images.Image{
	Name:   "docker.io/library/ubuntu:latest"@" + digestFor(10).String(),
	Target: desc(10),
}
testImages := []images.Image{
		ubuntuLatest,
		ubuntuLatestDigested
...
```

```suggestion
			all:    []images.Image{ubuntuLatest, ubuntuLatestDigested},
```

File: daemon/containerd/image_delete.go [hunk 7]
```diff
@@ -53,123 +53,156 @@ import (
...
-	singleRef, err := i.isSingleReference(ctx, img)
-	if err != nil {
-		return nil, err
-	}
-	if !singleRef {
-		err := i.client.ImageService().Delete(ctx, img.Name)
+	var imgID image.ID
+	if img == nil {
+		if len(all) == 0 {
+			return []imagetypes.DeleteResponse{}, nil
```
Comment (vvoland) [inline daemon/containerd/image_delete.go:69]: This should be unreachable, as `resolveAllReferences` would return `ErrImageDoesNotExist`? Shouldn't this also return `ErrImageDoesNotExist` to match that behavior?

https://github.com/moby/moby/blob/04526ea0fe85acdf99cf6d77bf065c72892a641e/daemon/containerd/image_delete_test.go#L30-L33

File: daemon/containerd/image_delete.go [hunk 8]
```diff
@@ -205,19 +238,68 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
...
+		named = repoRef
+		if tagged, ok := named.(reference.Tagged); ok {
+			tag = tagged.Tag()
+		}
 	}
-	for _, ref := range refs {
-		if !isDanglingImage(ref) && ref.Name != img.Name {
-			return false, nil
+	for _, ref := range imgs {
+		if !isDanglingImage(ref) {
```
Comment (vvoland) [inline daemon/containerd/image_delete.go:268]: This assumes that all `imgs` have the same `Target`, right? Should we mention that in the Godoc (or check that at the beginning and return error if the assumption is broken), or it's highly unlikely that this function might be used in some other case?

File: daemon/containerd/image.go [hunk 9]
```diff
@@ -341,3 +339,131 @@ func (i *ImageService) getAllImagesWithRepository(ctx context.Context, ref refer
...
+				if cimg.Target.Digest != dgst {
+					// Ambiguous image reference, use reference name
+					log.G(ctx).WithField("image", refOrID).WithField("target", cimg.Target.Digest).Warn("digest reference points to image with a different digest")
+					dgst = cimg.Target.Digest
+				}
+			} else if !cerrdefs.IsNotFound(err) {
+				return nil, nil, convertError(err)
+			} else {
+				dgst = d.Digest()
+			}
```
Comment (vvoland) [inline daemon/containerd/image.go:372]: (just a note when reviewing, wasn't sure which cases exactly would hit that path)

For the `^(sha256:)?([a-f0-9]{4,64})$` there are few distinct reference cases (`refOrID`):
1. `sha256:abcdef7890123456789012345678901234567890123456789012345678901234` (algo prefixed full digest)
2. `sha256:abce` (algo prefixed truncated digest)
3. `abcdef7890123456789012345678901234567890123456789012345678901234` (unprefixed full digest)
4. `abce` (unprefixed truncated digest)

This is what the reference is resolved to in each case (`reference.ParseAnyReference`):
1. `docker.io/library/sha256:abcdef7890123456789012345678901234567890123456789012345678901234`
  type: reference.taggedReference
2. `docker.io/library/sha256:abce`
  type: reference.taggedReference
3. `sha256:abcdef7890123456789012345678901234567890123456789012345678901234`
  type: reference.digestReference
4. `docker.io/library/abce`
  type: reference.repository


(https://go.dev/play/p/FU-Ysh-I2cI)

**Review by vvoland — COMMENTED** [review]
Looks good overall. Just a few comments/questions from me. Don't think any of those is a blocker though.

**Comment by vvoland** [comment]
Oh, also noticed this breaks the `TestRemoveByDigest` test.
The thing it checks is that you should be able to delete the image using the named and digested reference (taken from `RepoDigests`).

```bash
$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
busybox      latest    3fbc63216742   4 months ago   6.09MB

$ docker inspect busybox -f {{.RepoDigests}}
[busybox@sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79]

# This should work
$ docker rmi busybox@sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79
Error response from daemon: No such image: busybox@sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79
```

File: daemon/containerd/image.go (hunk 9 above)
Comment (dmcgowan) [reply daemon/containerd/image.go:372]: 1 Seems a bit surprising. I can add more unit tests around these cases to lessen the surprise.

**Comment by dmcgowan** [comment]
@vvoland I'll add more unit tests around that case. I don't think the current tests match all the format of references which get created.

File: daemon/containerd/image_delete.go (hunk 7 above)
Comment (dmcgowan) [reply daemon/containerd/image_delete.go:69]: It definitely should be unreachable but if it is an error is easier to follow than a panic or no-op.

File: daemon/containerd/image.go [hunk 10]
```diff
@@ -341,3 +339,131 @@ func (i *ImageService) getAllImagesWithRepository(ctx context.Context, ref refer
...
+		}
+
+		name := reference.TagNameOnly(named).String()
+
+		cimg, err := i.images.Get(ctx, name)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, nil, convertError(err)
+			}
+			return nil, nil, images.ErrImageDoesNotExist{Ref: parsed}
```
Comment (dmcgowan) [inline daemon/containerd/image.go:426]: Related to the remove by repo digest test, this is the case that gets hit. There is currently a unit test that checks for this condition. When a digest is provided we can probably fall back to similar logic as when just a hash is provided. The tricky part is whether to return a matched image or not, since in this case there was no actual matched image and trying to avoid any sort of fuzzy matching with whats in containerd. However I think it makes sense to support the `repo@digest` case even if it is new functionality and doesn't represent any specific image which has been created.

**Comment by dmcgowan** [comment]
Updated with functionality to match images by repo + digest on rmi.

**Comment by rumpl** [comment]
A lot of tests are failing now in integration-cli with `assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest: [d411442a41e1e] failed to download busybox` https://github.com/moby/moby/actions/runs/7177758758?pr=46840#summary-19545421070

**Comment by dmcgowan** [comment]
Updated, @rumpl late night commit, I missed the zero case. I added a unit test for that now as well.

Still looking at TestDockerCLIRmiSuite/TestRmiUntagHistoryLayer  but that isn't holding this PR back.

File: daemon/containerd/image.go [hunk 11]
```diff
@@ -352,3 +350,132 @@ func imageFamiliarName(img containerdimages.Image) string {
...
+		for _, rimg := range imgs {
+			if rimg.Name == img.Name {
+				err = nil
+				break
+			}
+		}
+	}
+	if errors.Is(err, errInconsistentData) {
+		if retries, ok := ctx.Value(errInconsistentData).(int); !ok || retries < 3 {
+			log.G(ctx).WithField("retry", retries).WithField("ref", refOrID).Info("image changed during lookup, retrying")
```
Comment (vvoland) [inline daemon/containerd/image.go:474]: nit
```suggestion
			log.G(ctx).WithFields(log.Fields{"retry": retries, "ref": refOrID}).Info("image changed during lookup, retrying")
```

File: daemon/containerd/service.go [hunk 12]
```diff
@@ -6,15 +6,17 @@ import (
...
+	"github.com/containerd/containerd/images"
 	"github.com/containerd/containerd/plugin"
 	"github.com/containerd/containerd/remotes/docker"
 	"github.com/containerd/containerd/snapshots"
 	"github.com/containerd/log"
 	"github.com/distribution/reference"
 	"github.com/docker/docker/container"
 	daemonevents "github.com/docker/docker/daemon/events"
-	"github.com/docker/docker/daemon/images"
+	daemonimages "github.com/docker/docker/daemon/images"
```
Comment (vvoland) [inline daemon/containerd/service.go:19]: Let's keep the aliases consistent:
```suggestion
	dimages "github.com/docker/docker/daemon/images"
```

File: daemon/containerd/image_delete_test.go [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+	"context"
+	"testing"
+
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/metadata"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/log/logtest"
+	"github.com/docker/docker/container"
+	daemonevents "github.com/docker/docker/daemon/events"
+	dockerimages "github.com/docker/docker/daemon/images"
```
Comment (vvoland) [inline daemon/containerd/image_delete_test.go:13]: Let's keep the aliases consistent:
```suggestion
	dimages "github.com/docker/docker/daemon/images"
```

File: daemon/containerd/image_test.go [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+	"math/rand"
+	"path/filepath"
+	"testing"
+
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/metadata"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/log/logtest"
+	"github.com/distribution/reference"
+	dockerimages "github.com/docker/docker/daemon/images"
```
Comment (vvoland) [inline daemon/containerd/image_test.go:15]: Let's keep the aliases consistent:
```suggestion
	dimages "github.com/docker/docker/daemon/images"
```

File: daemon/containerd/image_delete.go [hunk 15]
```diff
@@ -53,123 +54,180 @@ import (
...
-	if !singleRef {
-		err := i.client.ImageService().Delete(ctx, img.Name)
+	var imgID image.ID
+	if img == nil {
+		if len(all) == 0 {
+			return []imagetypes.DeleteResponse{}, nil
+		}
+		imgID = image.ID(all[0].Target.Digest)
+		var named reference.Named
+		if imageRef != string(imgID) {
```
Comment (vvoland) [inline daemon/containerd/image_delete.go:74]: Shouldn't this also check for truncated ID (and the dangling image)?
Like here:
https://github.com/moby/moby/blob/5c48d187eeb356f5617a3ffda5adfcfef5d5b986/daemon/containerd/image_delete.go#L107-L108

If so, we should probably extract that to a function an call it here.

File: daemon/containerd/image.go [hunk 16]
```diff
@@ -352,3 +350,132 @@ func imageFamiliarName(img containerdimages.Image) string {
...
 }
+
+func convertError(err error) error {
+	// TODO: Convert containerd error to Docker error
+	return err
+}
+
+// resolveAllReferences resolves the reference name or ID to an image and returns all the images with
+// the same target.
+func (i *ImageService) resolveAllReferences(ctx context.Context, refOrID string) (*containerdimages.Image, []containerdimages.Image, error) {
```
Comment (vvoland) [inline daemon/containerd/image.go:361]: I think it would be worth to highlight that the function can return nil as the first argument (`*containerdimages.Image`) while returning a non empty image slice (`[]containerdimages.Image`) and a nil error. This can only happen if the `refOrID` doesn't match to any image name, only the image target digest?

File: daemon/containerd/image_delete.go [hunk 17]
```diff
@@ -53,123 +54,180 @@ import (
...
+		if named != nil && len(sameRef) > 0 && len(sameRef) != len(all) {
+			familiarRef := reference.FamiliarString(named)
+			for _, ref := range sameRef {
+				// TODO: Add with target
+				err := i.images.Delete(ctx, ref.Name)
+				if err != nil {
+					return nil, err
+				}
 			}
+			i.logImageEvent(sameRef[0], familiarRef, events.ActionUnTag)
```
Comment (vvoland) [inline daemon/containerd/image_delete.go:101]: Shouldn't the event be logged for every ref from `sameRef`?

File: daemon/containerd/image_delete.go [hunk 18]
```diff
@@ -53,123 +54,180 @@ import (
...
 		}
+		if len(sameRef) != len(all) {
+			for _, ref := range sameRef {
+				// TODO: Add with target
+				err := i.images.Delete(ctx, ref.Name)
+				if err != nil {
+					return nil, err
+				}
+			}
+			i.logImageEvent(*img, familiarRef, events.ActionUnTag)
```
Comment (vvoland) [inline daemon/containerd/image_delete.go:129]: Shouldn't the event be logged for every ref from `sameRef`?

File: daemon/containerd/image_delete.go (hunk 15 above)
Comment (dmcgowan) [reply daemon/containerd/image_delete.go:74]: Could check not a prefix, no need to check dangling since it would already hit this branch as not a hash or hash prefix.

Comment (dmcgowan) [reply daemon/containerd/image_delete.go:74]: Actually just `!isImageIDPrefix(imgID.String(), imageRef)` is enough here since that covers the equal case already

File: daemon/containerd/image_delete.go (hunk 17 above)
Comment (dmcgowan) [reply daemon/containerd/image_delete.go]: Probably, I'll update and see if any of the tests complain. Either I was worried that the log output would differ here or wasn't quite sure that it would matter.

File: daemon/containerd/image.go (hunk 16 above)
Comment (dmcgowan) [reply daemon/containerd/image.go]: Added a longer section explaining the output. Gofmt seemed to really want me to go big with the formatting but think it is clearer now.

File: daemon/containerd/image.go [hunk 19]
```diff
@@ -352,3 +350,151 @@ func imageFamiliarName(img containerdimages.Image) string {
...
+
+		cimg, err := i.images.Get(ctx, name)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, nil, convertError(err)
+			}
+			if dgst == "" {
+				return nil, nil, images.ErrImageDoesNotExist{Ref: parsed}
+			}
+		} else {
```
Comment (thaJeztah) [inline daemon/containerd/image.go:459]: Just a quick drive-by comment (didn't look at the whole logic); here we are _NOT_ returning for a `NotFound` error if `dgst == ""`, correct?

If that's intentional, it may be worth putting a comment in that if-branch

Comment (dmcgowan) [reply daemon/containerd/image.go]: Added comment

**Comment by dmcgowan** [comment]
Updated with comment and added a commit which fixes `TestRmiUntagHistoryLayer`.

The fix for the untag is to re-add the dangling image before deleting the last tag when there are children and the rmi is not forced. I think this matches the intended logic around dangling image, another label scan right before delete is not optimal but we might be able to address that with switching to gc labels in 2.0.

File: daemon/containerd/image_delete.go [hunk 20]
```diff
@@ -5,13 +5,16 @@ import (
 	"fmt"
 	"sort"
 	"strings"
+	"time"
 
+	"github.com/containerd/containerd/errdefs"
```
Comment (vvoland) [inline daemon/containerd/image_delete.go:10]: Linter is unhappy:
>  daemon/containerd/image_delete.go:10:2: import "github.com/containerd/containerd/errdefs" imported without alias but must be with alias "cerrdefs" according to config (importas)
	"github.com/containerd/containerd/errdefs"
	^

```suggestion
	cerrdefs "github.com/containerd/containerd/errdefs"
```

**Review by vvoland — APPROVED** [review]
LGTM, thanks!

**Review by thaJeztah — APPROVED** [review]
LGTM

File: integration-cli/docker_cli_rmi_test.go [hunk 21]
```diff
@@ -295,7 +295,7 @@ RUN echo 2 #layer2
 	// should not be untagged without the -f flag
 	assert.ErrorContains(c, err, "")
 	assert.Assert(c, strings.Contains(out, cID[:12]))
-	assert.Assert(c, strings.Contains(out, "(must force)"))
+	assert.Assert(c, strings.Contains(out, "(must force)") || strings.Contains(out, "(must be forced)"))
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply integration-cli/docker_cli_rmi_test.go:298]: Oh! My search only found one, but perhaps I did it wrong.

Yeah, they were somewhat weird error messages in either case.
